server/routerlicious/packages/services-core/src/runWithRetry.ts

```diff
@@ -21,18 +34,18 @@ import { NetworkError } from "@fluidframework/server-services-client";
  * @param onErrorFn - function allowing caller to define custom logic to run on error e.g. custom logs
  * @param telemetryEnabled - whether to log telemetry metric, default is false
  */
-export async function runWithRetry<T>(
-    api: () => Promise<T>,
-    callName: string,
-    maxRetries: number,
-    retryAfterMs: number,
-    telemetryProperties?: Map<string, any> | Record<string, any>,
-    shouldIgnoreError?: (error) => boolean,
-    shouldRetry?: (error) => boolean,
+export async function runWithRetry<T>({
```

Comment:
Do you mean clients can directly be calling "runWithRetry" from outside?
Currently I see it being called in routerlicious and historian only(I searched in the whole repo), but if there can be more usage outside of this, I agree with your suggestion.

Reply:
As far as I know, the client does not use this `runWithRetry` definition. It uses [runWithRetry from driver-utils](https://github.com/microsoft/FluidFramework/blob/main/packages/loader/driver-utils/src/runWithRetry.ts).

Reply:
Maybe @zhangxin511 meant that customers of this OSS repo who implement their own Fluid service and who might consume this services-core package would see breaking changes from this?

Reply:
@znewton , @hedasilv Thx for the info. Yes, @hedasilv 's comments is what I mean here.

So we have existing api `X`, instead of change it to `X1` directly and introduce breaking changes, maybe we could just expose a new `X1`, and call `X` internally and mark `X` as obsolete, say will remove after 1 year. And then eventually just remove the older version `X`. Is this an option here in Fluid OSS package?

Reply:
Fair point. Currently we don't really do much to maintain compatibility for APIs in the R11s packages unless they are used internally (T9s and R11s-Driver). Even then, the best practice is just making sure that we bump the service package version so that T9s and R11s-driver don't accidentally pull it in.

I agree that we should probably be better about that, and we should probably help prevent this change from accidentally breaking for consumers of service packages. For this specific instance, we could use function overrides with a small check in the front of the function, with 2 runWithRetry internal function types, then stop adding functionality to the deprecated one. Something like

```ts
export async function runWithRetry<T>(
    api: () => Promise<T>,
    callName: string,
    maxRetries: number,
    retryAfterMs: number,
    telemetryProperties?: Map<string, any> | Record<string, any>,
    shouldIgnoreError?: (error) => boolean,
    shouldRetry?: (error) => boolean,
    calculateIntervalMs = (error, numRetries, retryAfterInterval) => retryAfterInterval * 2 ** numRetries,
    onErrorFn?: (error) => void,
    telemetryEnabled = false,
): Promise<T>;
export async function runWithRetry<T>(options: IRunWithRetryOptions<T>): Promise<T>;
export async function runWithRetry<T>(
    apiOrOptions:  IRunWithRetryOptions() => Promise<T>,
    callName: string,
    maxRetries: number,
    retryAfterMs: number,
    telemetryProperties?: Map<string, any> | Record<string, any>,
    shouldIgnoreError?: (error) => boolean,
    shouldRetry?: (error) => boolean,
    calculateIntervalMs = (error, numRetries, retryAfterInterval) => retryAfterInterval * 2 ** numRetries,
    onErrorFn?: (error) => void,
    telemetryEnabled = false,
): Promise<T> {
  if (typeof apiOrOptions === "function") {
    return deprecatedRunWithRetry(...);
  } else {
    return modernRunWithRetry(apiOrOptions);
  }
}
```

Reply:
I'm a big fan of named params, but this feels like a lot of work to add on top of this PR, especially since this PR isn't actually changing any internal functionality of `runWithRetry`.